exporter/awsemfexporter/pusher.go

```diff
@@ -0,0 +1,313 @@
+	logEventBatch.sortLogEvents()
+	putLogEventsInput := logEventBatch.PutLogEventsInput
+
+	if p.streamToken == "" {
+		//log part and retry logic are already done inside the CreateStream
+		// when the error is not nil, the stream token is "", which is handled in the below logic.
+		p.streamToken, _ = p.svcStructuredLog.CreateStream(p.logGroupName, p.logStreamName)
+	}
+
+	if p.streamToken != "" {
+		putLogEventsInput.SequenceToken = aws.String(p.streamToken)
+	}
+
+	startTime := time.Now()
+
+	var tmpToken *string
+	var err error
+	tmpToken, err = p.svcStructuredLog.PutLogEvents(putLogEventsInput, p.retryCnt)
+
+	if err != nil {
+		return err
+	}
+
+	log.Printf("D! logpusher: publish %d log events with size %f KB in %d ms.",
+		len(putLogEventsInput.LogEvents),
+		float64(logEventBatch.byteTotal)/float64(1024),
+		time.Since(startTime).Nanoseconds()/int64(time.Millisecond))
+
+	if tmpToken != nil {
+		p.streamToken = *tmpToken
+	}
+	diff := time.Since(startTime)
+	if timeLeft := minPusherIntervalInMillis*time.Millisecond - diff; timeLeft > 0 {
+		time.Sleep(timeLeft)
+	}
+	return nil
+}
+
+//Create a new log event batch if needed.
+func (p *pusher) newLogEventBatch() *LogEventBatch {
+	logEventBatch := &LogEventBatch{
+		PutLogEventsInput: &cloudwatchlogs.PutLogEventsInput{
+			LogGroupName:  p.logGroupName,
+			LogStreamName: p.logStreamName,
+			LogEvents:     make([]*cloudwatchlogs.InputLogEvent, 0, MaxRequestEventCount)},
+		creationTime: time.Now(),
+	}
+	return logEventBatch
+}
+
+//Determine if a new log event batch is needed.
+func (p *pusher) newLogEventBatchIfNeeded(logEvent *LogEvent) error {
+	var err error
+	logEventBatch := p.logEventBatch
+	if len(logEventBatch.PutLogEventsInput.LogEvents) == cap(logEventBatch.PutLogEventsInput.LogEvents) ||
+		logEvent != nil && (logEventBatch.byteTotal+logEvent.eventPayloadBytes() > MaxRequestPayloadBytes || !logEventBatch.timestampWithin24Hours(logEvent.InputLogEvent.Timestamp)) {
+		err = p.pushLogEventBatch(logEventBatch)
+		p.logEventBatch = p.newLogEventBatch()
+	}
+	return err
+}
+
+func (p *pusher) flushLogEventBatch() error {
+	var err error
+	if len(p.logEventBatch.PutLogEventsInput.LogEvents) > 0 {
+		logEventBatch := p.logEventBatch
+		err = p.pushLogEventBatch(logEventBatch)
+		p.logEventBatch = p.newLogEventBatch()
+	}
+	return err
+}
+
+//Add the log event onto the log event batch
+func (p *pusher) addLogEvent(logEvent *LogEvent) error {
+	var err error
+	if len(*logEvent.InputLogEvent.Message) == 0 {
+		return nil
+	}
+
+	//http://docs.aws.amazon.com/goto/SdkForGoV1/logs-2014-03-28/PutLogEvents
+	//* None of the log events in the batch can be more than 2 hours in the
+	//future.
+	//* None of the log events in the batch can be older than 14 days or the
+	//retention period of the log group.
+	currentTime := time.Now().UTC()
+	utcTime := time.Unix(0, *logEvent.InputLogEvent.Timestamp*int64(time.Millisecond)).UTC()
+	duration := currentTime.Sub(utcTime).Hours()
```

Comment:
Is it ok to round the hours? How about converting `24*14` to duration instead?

Reply:
Should probably have constants for the durations for the staleness check

Reply:
Agree it's better to use constants here.